src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -41,7 +41,32 @@ private static void RootReflectionSerializerDependencies()
             }
 
             [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-            static JsonTypeInfo CreateJsonTypeInfo(Type type, JsonSerializerOptions options) => new JsonTypeInfo(type, options);
+            static JsonTypeInfo CreateJsonTypeInfo(Type type, JsonSerializerOptions options)
+            {
+                JsonTypeInfo.ValidateType(type, null, null, options);
+
+                MethodInfo methodInfo = typeof(JsonSerializerOptions).GetMethod(nameof(CreateReflectionJsonTypeInfo), BindingFlags.NonPublic | BindingFlags.Instance)!;
+                // Some of the validation is done during construction (i.e. validity of JsonConverter, inner types etc.)
+                // therefore we need to unwrap TargetInvocationException for better user experience
+#if NETCOREAPP
+                return (JsonTypeInfo)methodInfo.MakeGenericMethod(type).Invoke(options, BindingFlags.NonPublic | BindingFlags.DoNotWrapExceptions, null, null, null)!;
+#else
+                try
+                {
+                    return (JsonTypeInfo)methodInfo.MakeGenericMethod(type).Invoke(options, null)!;
+                }
+                catch (TargetInvocationException ex)
+                {
+                    throw ex.InnerException!;
```

Comment:
Related to my comment in line 49, but additionally calling `throw ex.InnerException;` will reset the stacktrace. If we absolutely need to do this, you might want to do `ExceptionDispatchInfo.Capture(ex.InnerException).Throw();` here.